src/tcache.c

```diff
@@ -390,40 +426,68 @@ tcache_event(tsd_t *tsd) {
 	}
 
 	tcache_slow_t *tcache_slow = tsd_tcache_slowp_get(tsd);
-	szind_t szind = tcache_slow->next_gc_bin;
-	bool is_small = (szind < SC_NBINS);
-	cache_bin_t *cache_bin = &tcache->bins[szind];
+	assert(tcache_slow != NULL);
 
-	if (tcache_bin_disabled(szind, cache_bin, tcache_slow)) {
-		goto label_done;
+	/* When the new tcache gc is not enabled, GC one bin at a time. */
+	if (!opt_experimental_tcache_gc) {
+		szind_t szind = tcache_slow->next_gc_bin;
+		bool is_small = (szind < SC_NBINS);
+		tcache_gc(tsd, tcache_slow, tcache, szind, is_small);
+		tcache_slow->next_gc_bin++;
+		if (tcache_slow->next_gc_bin == tcache_nbins_get(tcache_slow)) {
+			tcache_slow->next_gc_bin = 0;
+		}
+		return;
 	}
 
-	tcache_bin_flush_stashed(tsd, tcache, cache_bin, szind, is_small);
-	cache_bin_sz_t low_water = cache_bin_low_water_get(cache_bin);
-	if (low_water > 0) {
-		if (is_small) {
-			tcache_gc_small(tsd, tcache_slow, tcache, szind);
-		} else {
-			tcache_gc_large(tsd, tcache_slow, tcache, szind);
+	nstime_t now;
+	nstime_init_update(&now);
+	/* Non monotonic clock, time went backward, calibrate the last gc time. */
+	if (unlikely(nstime_compare(&tcache_slow->last_gc_time, &now) > 0)) {
+		assert(!nstime_monotonic());
+		nstime_copy(&tcache_slow->last_gc_time, &now);
+	}
+	assert(nstime_compare(&tcache_slow->last_gc_time, &now) <= 0);
+	uint64_t interval = (uint64_t)10 * KQU(1000000);	/* 10ms */
+	if (now.ns - tcache_slow->last_gc_time.ns < interval) {
+		// time interval is too short, skip this event.
+		return;
+	}
+	/* Update last_gc_time to now. */
+	nstime_copy(&tcache_slow->last_gc_time, &now);
+
+	unsigned gc_small_nbins = 0;
+	unsigned tcache_nbins = tcache_nbins_get(tcache_slow);
+	unsigned small_nbins = tcache_nbins > SC_NBINS ? SC_NBINS : tcache_nbins;
+	szind_t ind_gc_small = tcache_slow->next_gc_bin_small;
+	szind_t ind_gc_large = tcache_slow->next_gc_bin_large;
+
+	for (unsigned i = 0; i < small_nbins; i++) {
+		if (ind_gc_small == small_nbins) {
+			ind_gc_small = 0;
 		}
-	} else if (is_small && tcache_slow->bin_refilled[szind]) {
-		assert(low_water == 0);
-		/*
-		 * Increase fill count by 2X for small bins.  Make sure
-		 * lg_fill_div stays greater than 0.
-		 */
-		if (tcache_slow->lg_fill_div[szind] > 1) {
-			tcache_slow->lg_fill_div[szind]--;
+		assert(ind_gc_small < SC_NBINS);
+		if ((gc_small_nbins += tcache_gc(tsd, tcache_slow, tcache,
```

Comment:
tcache_gc returns a bool. Let's break this into separate lines. Maybe:
```
if (tcache_gc(...)) {
    gc_small_nbins++;
}
```
and then move the TCACHE_GC_SMALL_NBINS_MAX check to the for loop 
`i < small_nbins && gc_small_nbins < TCACHE_GC_SMALL_NBINS_MAX`